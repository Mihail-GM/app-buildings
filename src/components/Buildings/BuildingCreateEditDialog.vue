<template>
	<v-dialog
		v-model="dialog"
		max-width="600px"
		@click:outside=close
	>
		<template v-slot:activator="{ on, attrs }">
			<v-btn
				color="primary"
				dark
				class="mb-2 px-8"
				:color="'green'"
				v-bind="attrs"
				v-on="on"
			>
				Add building
			</v-btn>
		</template>

		<v-card>
			<v-card-title class="d-flex justify-center">
				<span id="dialog-title" class="text-h5 text-center">{{ formTitle }}</span>
			</v-card-title>

			<v-card-text>
				<v-container>
					<v-row>
						<v-col
							cols="12"
							sm="12"
							md="12"
							class="py-0"
						>
							<v-text-field
								:disabled="!isCreateTable"
								v-model="editedItem.id"
								dense
								label="Building id"
								outlined
								required
								number
								:error-messages="buildingIdError"
								@blur="$v.editedItem.id.$touch()"
							/>

							<v-text-field
								v-model="editedItem.name"
								dense
								label="Building name"
								outlined
								required
								:error-messages="buildingNameError"
								@blur="$v.editedItem.name.$touch()"
							/>

							<v-text-field
								v-model="editedItem.area"
								dense
								label="Building area (m2)"
								outlined
								required
								type="number"
								:error-messages="buildingAreaError"
								@blur="$v.editedItem.area.$touch()"
							/>

							<v-text-field
								v-model="editedItem.location"
								dense
								label="Building location"
								outlined
							/>

							<v-select
								v-model="editedItem.type"
								dense
								label="Building type"
								:items="types"
								item-text="name"
								item-value="value"
								outlined
							/>
						</v-col>
					</v-row>
				</v-container>
			</v-card-text>

			<v-card-actions>
				<v-spacer></v-spacer>
				<v-btn
					id="cancel-button"
					color="red darken-1"
					style="color:white"
					@click.native="close"
					text
				>
					Cancel
				</v-btn>

				<v-btn
					id="create-edit-button"
					color="green darken-1"
					style="color:white"
					@click.native="save"
					:disabled="$v.$anyError || buildingIdError.length !== 0"
					text
				>
					{{ buttonText }}
				</v-btn>
			</v-card-actions>
		</v-card>
	</v-dialog>
</template>

<script>
	import {validationMixin} from 'vuelidate'
	import {required, minValue} from 'vuelidate/lib/validators'
	import {mapGetters} from "vuex";

	export default {
		name: "BuildingCreateEditDialog",

		components: {},

		props: {
			value: {
				Boolean,
				default: false
			},

			editedItemProp: {
				type: Object,
				required: true,
				default: {
					id: null,
					name: '',
					area: '',
					location: '',
					type: '',
					img: '',
				}
			},

			editedIndexProp: {
				type: Number,
				required: true,
				default: false
			},
		},

		mixins: [validationMixin],

		validations: {
			editedItem: {
				id: {required, minValue: minValue(1)},
				name: {required},
				area: {required, minValue: minValue(1)},
			}
		},

		data() {
			return {
				complete: false,
				editedItem: this.editedItemProp,
				editedIndex: this.editedIndexProp,

				defaultItem: {
					id: null,
					name: '',
					area: null,
					location: '',
					type: null,
					img: '',
				},

				types: [
					{
						value: 'residential',
						name: 'Residential'
					},
					{
						value: 'industrial',
						name: 'Industrial'
					},
					{
						value: 'educational',
						name: 'Educational'
					},
					{
						value: 'special',
						name: 'Special'
					},
				]
			}
		},

		methods: {
			close() {
				this.dialog = false;
				this.$emit('resetEditedItem');
				this.editedIndex = -1;
				this.editedItem = Object.assign({}, this.defaultItem);
				this.$v.$reset();
			},

			save() {
				this.$v.$touch();

				if (this.$v.$anyError) {
					this.$store.dispatch('snackbar/showSnackbarMessage', {
						message: 'Please enter valid data',
						duration: 6000,
						mode: 'fail'
					});
					return
				}

				this.$emit('save', this.editedItem);
				this.close();
			},
		},

		computed: {
			...mapGetters(
				{
					buildings: 'buildings/getBuildings',
				}
			),

			isCreateTable() {
				return this.editedIndex === -1;
			},

			formTitle() {
				return this.isCreateTable ? 'Create new building' : 'Edit building';
			},

			buttonText() {
				return this.isCreateTable ? 'CREATE PROJECT' : 'EDIT PROJECT';
			},

			isIdUniqueRequired() {
				return this.isCreateTable && this.buildings.some(oneBuilding => oneBuilding.id === this.editedItem.id)
			},

			dialog: {
				get() {
					return this.value
				},

				set(value) {
					this.$emit('input', value)
				}
			},

			buildingIdError() {
				const errors = []
				if (!this.$v.editedItem.id.$dirty) {
					return errors
				}

				!this.$v.editedItem.id.minValue && errors.push('Enter positive number for id')
				!this.$v.editedItem.id.required && errors.push('Building id is required.')
				this.isIdUniqueRequired && errors.push('Building id needs to be unique.')

				return errors
			},

			buildingNameError() {
				const errors = []
				if (!this.$v.editedItem.name.$dirty) {
					return errors
				}

				!this.$v.editedItem.name.required && errors.push('Building name is required.')
				return errors
			},

			buildingAreaError() {
				const errors = []
				if (!this.$v.editedItem.area.$dirty) {
					return errors
				}

				!this.$v.editedItem.area.minValue && errors.push('Enter positive number for area')
				!this.$v.editedItem.area.required && errors.push('Building area is required.')
				return errors
			},
		},
	}
</script>

<style lang="scss" scoped>
	.v-label {
		font-size: 14px !important;
	}

	.stripe-card {
		width: 100%;
		height: 45px;
		border-radius: 4px;
	}
</style>