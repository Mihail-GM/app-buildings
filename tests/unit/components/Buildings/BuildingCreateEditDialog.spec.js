// test/BuildingCreateEditDialog.spec.js

import Vuetify from 'vuetify'
import Vuex from "vuex";

import {createLocalVue, mount} from '@vue/test-utils'
import {merge} from "lodash";

import BuildingCreateEditDialog from "@/components/Buildings/BuildingCreateEditDialog";
import buildingsStore from "../../../../src/store/modules/buildings";
import flushPromises from "flush-promises";

describe('BuildingCreateEditDialog.vue', () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);
	let vuetify;
	let getters;

	beforeEach(() => {
		vuetify = new Vuetify();
	})

	function createStore(overrides) {
		let currentState = "DATA_LOADED";
		let buildings = [
			{
				area: "55",
				id: "2",
				img: "",
				location: "Sofia",
				name: "Luxury house 3",
				type: "house",
				uuid: "-MukqrOCyJiAZW1BCwI0",
			},
			{
				area: "55",
				id: "2",
				img: "",
				location: "Sofia",
				name: "Luxury house 3",
				type: "house",
				uuid: "-MukqrOCyJiAZW1BCwI0",
			}
		];

		getters = {
			getBuildings: () => currentState,
			getCurrentState: () => buildings
		}

		const defaultStoreConfig = {
			modules: {
				buildings: {
					namespaced: true,
					state: {
						currentState,
						buildings,
					},
					getters: buildingsStore.getters,

				}
			}
		};

		return new Vuex.Store(merge(defaultStoreConfig, overrides));
	}

	const createWrapper = (overrides) => {
		const defaultMountingOptions = {
			localVue,
			vuetify,
			store: createStore(),

			mocks: {
				// Vue Router
				$router: {
					push: () => {
					}
				},
			},
		};

		const propsData = {
			value: false,

			editedItemProp: {
				uuid: null,
				id: null,
				name: '',
				area: null,
				location: '',
				type: null,
				img: '',
			}
		}

		return mount(BuildingCreateEditDialog, merge(defaultMountingOptions, propsData, overrides))
	}

	it('should render component ', () => {
		const wrapper = createWrapper();

		expect(wrapper.is(BuildingCreateEditDialog)).toBe(true);
	})


	it('should update computed value when prop value changed', async () => {
		const wrapper = createWrapper();
		await wrapper.setProps({value: true});

		expect(wrapper.vm.dialog).toBe(true);
	})

	it('should emit input with false value when click close button', async () => {
		const closeChanged = jest.spyOn(BuildingCreateEditDialog.methods, "close");

		const wrapper = createWrapper();
		await wrapper.setProps({value: true});

		let closeButton = wrapper.find('#cancel-button');
		closeButton.trigger('click.stop');

		await flushPromises();
		expect(closeChanged).toHaveBeenCalled();
		expect(wrapper.emitted().input[0]).toEqual([false]);
	})

	it('should emit input with false value when confirm create', async () => {
		const closeChanged = jest.spyOn(BuildingCreateEditDialog.methods, "close");

		const wrapper = createWrapper();
		await wrapper.setProps({value: true});

		let confirmButton = wrapper.find('#create-edit-button');
		confirmButton.trigger('click.stop');

		await flushPromises();

		expect(closeChanged).toHaveBeenCalled();
	})

	it('should render text for edit dialog when we edit', async () => {
		const wrapper = createWrapper();
		await wrapper.setProps({value: true});
		await wrapper.setProps({editedIndexProp: 1});

		let dialogTitle = wrapper.find('#dialog-title');

		await flushPromises();

		expect(wrapper.vm.editedIndexProp).toBe(1);
		expect(dialogTitle.text()).toBe('Edit building');
	})

	it('should render text for create dialog when we add new buildings', async () => {
		const wrapper = createWrapper();
		await wrapper.setProps({value: true});
		await wrapper.setData({editedIndex: -1});

		let dialogTitle = wrapper.find('#dialog-title');

		expect(dialogTitle.text()).toBe('Create new building');
	})

	it('should not emit save event if data is not correct', async () => {
		const wrapper = createWrapper();

		await wrapper.setProps({value: true});
		await wrapper.setData({editedIndex: -1});
		await wrapper.setData({
			editedItem: {
				id: 5,
				name: null,
				area: null,
				location: 'Str. Address',
				type: 'apartment',
				img: '',
			}
		});
		await flushPromises();

		let confirmButton = wrapper.find('#create-edit-button');
		confirmButton.trigger('click.stop');
		await flushPromises();

		expect(wrapper.emitted().save).toBeFalsy();
	})

	it('should not emit save event if all data is not correct', async () => {
		const wrapper = createWrapper();

		await wrapper.setProps({value: true});
		await wrapper.setData({editedIndex: -1});
		await wrapper.setData({
			editedItem: {
				id: null,
				name: null,
				area: null,
				location: null,
				type: null,
				img: '',
			}
		});
		await flushPromises();

		let confirmButton = wrapper.find('#create-edit-button');
		confirmButton.trigger('click.stop');
		await flushPromises();

		expect(wrapper.emitted().save).toBeFalsy();
	})

	it('should emit save event if data is correct', async () => {
		const wrapper = createWrapper();

		await wrapper.setProps({value: true});
		await wrapper.setData({editedIndex: 0});
		await wrapper.setData({
			editedItem: {
				id: 5,
				name: 'Luxury building',
				area: '120',
				location: 'Str. Address',
				type: 'apartment',
				img: '',
			}
		});
		await flushPromises();

		let confirmButton = wrapper.find('#create-edit-button');
		confirmButton.trigger('click.stop');
		await flushPromises();

		expect(wrapper.emitted().save).toBeTruthy()
	})

	it('should match snapshot', () => {
		const wrapper = createWrapper()

		expect(wrapper.html()).toMatchSnapshot()
	})
})
