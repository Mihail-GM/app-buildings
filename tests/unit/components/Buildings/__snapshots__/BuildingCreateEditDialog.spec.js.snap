// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildingCreateEditDialog.vue should match snapshot 1`] = `
<div class="v-dialog__container"><button type="button" class="mb-2 px-8 v-btn v-btn--is-elevated v-btn--has-bg theme--dark v-size--default green" role="button" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content">
			Add building
		</span></button>
  <!---->
</div>
`;
